traffic_ops/app/lib/API/ApiCapability.pm

```diff
@@ -0,0 +1,258 @@
+
+	my $rs_data = $self->db->resultset("ApiCapability")->search( 'me.capability' => $capability );
+	$self->renderResults( $rs_data ) ;
+}
+
+sub index {
+	my $self = shift;
+	my $id = $self->param('id');
+
+	my $rs_data = $self->db->resultset("ApiCapability")->search( 'me.id' => $id );
+	$self->renderResults( $rs_data ) ;
+}
+
+sub is_mapping_valid {
+	my $self = shift;
+	my $id = shift;
+	my $http_method = shift;
+	my $route = shift;
+	my $capability = shift;
+
+	if ( !defined($http_method) ) {
+		return ( undef, "HTTP method is required." );
+	}
+
+	if ( !exists( $valid_http_methods{ $http_method } ) ) {
+		return ( undef, "HTTP method \'$http_method\' is invalid. Valid values are: " . join(", ", sort keys %valid_http_methods ) );
+	}
+
+	if ( !defined($route) or $route eq "" ) {
+		return ( undef, "Route is required." );
+	}
+
+	if ( !defined($capability) or $capability eq "" ) {
+		return (undef, "Capability name is required." );
+	}
+	# check if capability exists
+	my $rs_data = $self->db->resultset("Capability")->search( { 'name' => { 'like', $capability } } )->single();
+	if (!defined($rs_data)) {
+		return (undef, "Capability '$capability' does not exist." );
+	}
+
+	# search a mapping for the same http_method & route
+	$rs_data = $self->db->resultset("ApiCapability")->search( { 'route' => { 'like', $route } } )->search( {
+		'http_method' => { '=', $http_method } } )->single();
+	# if adding a new entry, make sure it is unique
+	if ( !defined( $id ) ) {
+		if (defined($rs_data)) {
+			my $allocated_capability = $rs_data->capability->name;
+			return (undef, "HTTP method '$http_method', route '$route' are already mapped to capability: $allocated_capability" );
+		}
+	}
+	else {
+		if (defined($rs_data)) {
+			my $lid = $rs_data->id;
+			if ($lid ne $id) {
+				my $allocated_capability = $rs_data->capability->name;
+				return (undef, "HTTP method '$http_method', route '$route' are already mapped to capability: $allocated_capability" );
+			}
+		}
+	}
+
+	return ( 1, undef );
+}
+
+sub create {
+	my $self = shift;
+	my $params = $self->req->json;
+
+	if ( !&is_oper($self) ) {
+		return $self->forbidden();
+	}
+
+	if ( !defined($params) ) {
+		return $self->alert("Parameters must be in JSON format.");
+	}
+
+	my $http_method = $params->{httpMethod} if defined($params->{httpMethod});
+	my $route = $params->{route} if defined($params->{route});
+	my $capability = $params->{capName} if defined($params->{capName});
+	my $id = undef;
+
+	my ( $is_valid, $errStr ) = $self->is_mapping_valid( $id, $http_method, $route, $capability );
+	if ( !$is_valid ) {
+		return $self->alert( $errStr );
+	}
+
+	my $values = {
+		id 			=> $self->db->resultset('ApiCapability')->get_column('id')->max() + 1,
+		http_method	=> $http_method,
+		route		=> $route,
+		capability	=> $capability
+	};
+
+	my $insert = $self->db->resultset('ApiCapability')->create($values);
+	my $rs = $insert->insert();
+	if ($rs) {
+		my $response;
+		$response->{id}				= $rs->id;
```

Comment:
it looks like you might have to run your code thru perltidy. http://trafficcontrol.apache.org/docs/latest/development/traffic_ops.html#perl-formatting-conventions